File: shared/truv.js
import fetch from 'node-fetch';
import { v4 as uuidv4 } from 'uuid';

export class TruvClient {
  constructor({ clientId, secret, apiUrl = 'https://prod.truv.com/v1/' }) {
    this.clientId = clientId;
    this.secret = secret;
    this.apiUrl = apiUrl;
    this.headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'X-Access-Client-Id': clientId,
      'X-Access-Secret': secret,
    };
  }

  async _request(method, endpoint, { json } = {}) {
    const url = this.apiUrl + endpoint;
    const start = performance.now();

    const opts = { method, headers: this.headers };
    if (json) opts.body = JSON.stringify(json);

    const response = await fetch(url, opts);
    const durationMs = Math.round((performance.now() - start) * 10) / 10;

    let data;
    const text = await response.text();
    try {
      data = text ? JSON.parse(text) : {};
    } catch {
      data = { raw: text };
    }

    console.log(`TRUV: ${method.toUpperCase()} ${url} — ${response.status} (${durationMs}ms)`);
    return { statusCode: response.status, data, durationMs, requestBody: json || null };
  }

  // --- Users API ---

  async createUser(overrides = {}) {
    const payload = {
      external_user_id: `qs-${uuidv4()}`, // Replace with your internal user/application ID
      first_name: 'John',
      last_name: 'Johnson',
      email: 'j.johnson@example.com',
      ...overrides,
    };
    return this._request('POST', 'users/', { json: payload });
  }

  async createUserBridgeToken(userId, productType) {
    const payload = {
      product_type: productType,
      client_name: 'Truv Quickstart',
      tracking_info: '1338-0111-A', // Replace with your internal reference (e.g. loan number)
    };

    // Sandbox test account for deposit_switch and pll products
    if (productType === 'deposit_switch' || productType === 'pll') {
      payload.account = {
        account_number: '16002600',   // Sandbox test value — replace with real account in production
        account_type: 'checking',
        routing_number: '12345678',   // Sandbox test value
        bank_name: 'Truv Bank',       // Sandbox test value
      };
      if (productType === 'pll') {
        payload.account.deposit_type = 'amount';
        payload.account.deposit_value = 100; // Numeric amount
      }
    }

    return this._request('POST', `users/${userId}/tokens/`, { json: payload });
  }

  // --- Company Search ---

  async searchCompanies(query, productType) {
    const params = new URLSearchParams({ query });
    if (productType) params.set('product_type', productType);
    return this._request('GET', `company-mappings-search/?${params}`);
  }

  // --- Orders API ---

  async createOrder(params = {}) {
    const productType = params.product_type || 'income';
    const payload = {
      order_number: `qs-${uuidv4()}`, // Replace with your internal order/application ID
      external_user_id: params.external_user_id || `qs-${uuidv4()}`,
      first_name: params.first_name || 'John',
      last_name: params.last_name || 'Johnson',
      products: params.products || [productType],
    };

    if (params.email) payload.email = params.email;
    if (params.phone) payload.phone = params.phone;
    if (params.ssn) payload.social_security_number = params.ssn;
    if (params.template_id) payload.template_id = params.template_id;

    // Employer — sandbox credentials: goodlogin/goodpassword
    if (params.company_mapping_id) {
      payload.employers = [{ company_mapping_id: params.company_mapping_id }];
    } else if (params.employer) {
      payload.employers = [{ company_name: params.employer }];
    }

    // Sandbox test account for deposit_switch and pll products
    if (['deposit_switch', 'pll'].includes(productType) && payload.employers) {
      payload.employers[0].account = {
        account_number: '16002600',   // Sandbox test value — replace with real account in production
        account_type: 'checking',
        routing_number: '12345678',   // Sandbox test value
        bank_name: 'Truv Bank',       // Sandbox test value
      };
      if (productType === 'pll') {
        payload.employers[0].account.deposit_type = 'amount';
        payload.employers[0].account.deposit_value = 100; // Numeric amount
      }
    }

    return this._request('POST', 'orders/', { json: payload });
  }

  async getOrder(truvOrderId) {
    return this._request('GET', `orders/${truvOrderId}/`);
  }

  async refreshOrder(truvOrderId) {
    return this._request('POST', `orders/${truvOrderId}/refresh/`);
  }

  // --- Webhooks API ---

  async listWebhooks() {
    return this._request('GET', 'webhooks/');
  }

  async createWebhook(params) {
    return this._request('POST', 'webhooks/', { json: params });
  }

  async deleteWebhook(webhookId) {
    return this._request('DELETE', `webhooks/${webhookId}/`);
  }

  // --- Token Exchange & Reports ---

  async getAccessToken(publicToken) {
    return this._request('POST', 'link-access-tokens/', { json: { public_token: publicToken } });
  }

  async getLinkReport(linkId, productType) {
    return this._request('GET', `links/${linkId}/${productType}/report`);
  }

  // --- Reports ---

  // VOIE/VOE report: is_voe=false → income+employment, is_voe=true → employment only
  async createVoieReport(userId, isVoe = false) {
    return this._request('POST', `users/${userId}/reports/`, { json: { is_voe: isVoe } });
  }

  async getVoieReport(userId, reportId) {
    return this._request('GET', `users/${userId}/reports/${reportId}/`);
  }

  // Assets report
  async createAssetsReport(userId) {
    return this._request('POST', `users/${userId}/assets/reports/`);
  }

  async getAssetsReport(userId, reportId) {
    return this._request('GET', `users/${userId}/assets/reports/${reportId}/`);
  }

  // Income insights report
  async createIncomeInsightsReport(userId) {
    return this._request('POST', `users/${userId}/income_insights/reports/`, {
      json: { days_requested: 60, consumer_report_permissible_purpose: 'EXTENSION_OF_CREDIT' },
    });
  }

  async getIncomeInsightsReport(userId, reportId) {
    return this._request('GET', `users/${userId}/income_insights/reports/${reportId}/`);
  }

  // Legacy
  async getVoaReport(userId, reportId) {
    return this._request('GET', `users/${userId}/assets/reports/${reportId}/`);
  }

  // --- Document Collections API ---

  async createDocumentCollection(documents, users) {
    return this._request('POST', 'documents/collections/', {
      json: { documents, ...(users ? { users } : {}) },
    });
  }

  async getDocumentCollection(collectionId) {
    return this._request('GET', `documents/collections/${collectionId}/`);
  }

  async uploadToCollection(collectionId, documents) {
    return this._request('POST', `documents/collections/${collectionId}/upload/`, {
      json: { documents },
    });
  }

  async finalizeCollection(collectionId) {
    return this._request('POST', `documents/collections/${collectionId}/finalize/`);
  }

  async getFinalizationResults(collectionId) {
    return this._request('GET', `documents/collections/${collectionId}/finalize/`);
  }
}

File: examples/python/src/truv.py
import logging
import time
from uuid import uuid4

import requests
from faker import Faker

fake = Faker()


class TruvClient:
    api_url = "https://prod.truv.com/v1/"

    def __init__(
        self, client_id: str, secret: str, product_type: str = "income", api_url: str = None
    ):
        self.client_id = client_id
        self.secret = secret
        self.headers = {
            "X-Access-Client-Id": client_id,
            "X-Access-Secret": secret,
            "Content-Type": "application/json;charset=UTF-8",
            "Accept": "application/json",
        }
        if api_url:
            self.api_url = api_url
        self.product_type = product_type

    def _request(self, method, endpoint, **kwargs) -> dict:
        """Make an API request and return {status_code, data, duration_ms}."""
        headers = kwargs.pop("headers", {})
        headers.update(self.headers)

        url = self.api_url + endpoint
        start = time.perf_counter()

        try:
            response = requests.request(
                method,
                url,
                headers=headers,
                timeout=15,
                **kwargs,
            )
            duration_ms = round((time.perf_counter() - start) * 1000, 1)

            logging.info(
                "TRUV: Response: %s %s - %s (%.1fms)",
                method.upper(),
                url,
                response.status_code,
                duration_ms,
            )

            data = None
            try:
                data = response.json()
            except Exception:
                data = {"raw": response.text}

            return {
                "status_code": response.status_code,
                "data": data,
                "duration_ms": duration_ms,
            }

        except requests.exceptions.RequestException as err:
            duration_ms = round((time.perf_counter() - start) * 1000, 1)
            logging.exception("API Request Error: %s", err)
            raise

    def post(self, endpoint: str, **kwargs) -> dict:
        return self._request("post", endpoint, **kwargs)

    def get(self, endpoint: str, **kwargs) -> dict:
        return self._request("get", endpoint, **kwargs)

    # --- Users API ---

    def create_user(self, **kwargs) -> dict:
        logging.info("TRUV: Creating new user")
        payload = {
            "external_user_id": f"qs-{uuid4().hex}",
            "first_name": fake.first_name(),
            "last_name": fake.last_name(),
            "email": fake.email(domain="example.com"),
            **kwargs,
        }
        return self.post("users/", json=payload)

    def create_user_bridge_token(self, user_id: str, product_type: str = None) -> dict:
        logging.info("TRUV: Creating bridge token for user %s", user_id)
        pt = product_type or self.product_type

        payload = {
            "product_type": pt,
            "tracking_info": "1338-0111-A",
        }

        if pt in ["deposit_switch", "pll"]:
            payload["account"] = {
                "account_number": "16002600",
                "account_type": "checking",
                "routing_number": "12345678",
                "bank_name": fake.company(),
            }
            if pt == "pll":
                payload["account"].update({
                    "deposit_type": "amount",
                    "deposit_value": "100",
                })

        return self.post(f"users/{user_id}/tokens/", json=payload)

    # --- Orders API ---

    def create_order(self, params: dict = None) -> dict:
        """POST /v1/orders/ — create an order (embedded or hosted)."""
        logging.info("TRUV: Creating order")
        if params is None:
            params = {}

        product_type = params.get("product_type", self.product_type)

        payload = {
            "order_number": f"qs-{uuid4().hex}",
            "first_name": params.get("first_name") or fake.first_name(),
            "last_name": params.get("last_name") or fake.last_name(),
            "email": params.get("email") or fake.email(domain="example.com"),
            "products": [product_type],
        }

        if params.get("phone"):
            payload["phone"] = params["phone"]

        if params.get("ssn"):
            payload["social_security_number"] = params["ssn"]

        if product_type in ["deposit_switch", "pll", "employment", "income"]:
            payload["employers"] = [{"company_name": "Home Depot"}]

        if product_type in ["deposit_switch", "pll"]:
            payload["employers"][0]["account"] = {
                "account_number": "16002600",
                "account_type": "checking",
                "routing_number": "12345678",
                "bank_name": "Truv Bank",
            }
            if product_type == "pll":
                payload["employers"][0]["account"].update({
                    "deposit_type": "amount",
                    "deposit_value": "100",
                })

        return self.post("orders/", json=payload)

    def get_order(self, truv_order_id: str) -> dict:
        """GET /v1/orders/{id}/ — get order details."""
        logging.info("TRUV: Getting order %s", truv_order_id)
        return self.get(f"orders/{truv_order_id}/")

    def refresh_order(self, truv_order_id: str) -> dict:
        """POST /v1/orders/{id}/refresh/ — refresh an order."""
        logging.info("TRUV: Refreshing order %s", truv_order_id)
        return self.post(f"orders/{truv_order_id}/refresh/")

    def get_order_certifications(self, truv_order_id: str) -> dict:
        """GET /v1/orders/{id}/certifications/ — get certifications."""
        logging.info("TRUV: Getting certifications for order %s", truv_order_id)
        return self.get(f"orders/{truv_order_id}/certifications/")

    # --- Token Exchange & Reports ---

    def get_access_token(self, public_token: str) -> dict:
        logging.info("TRUV: Exchanging public_token for access_token")
        return self.post("link-access-tokens/", json={"public_token": public_token})

    def get_link_report(self, link_id: str, product_type: str) -> dict:
        logging.info("TRUV: Getting %s report for link %s", product_type, link_id)
        return self.get(f"links/{link_id}/{product_type}/report")

    # --- Refresh Tasks (Users API) ---

    def create_refresh_task(self, access_token: str) -> dict:
        logging.info("TRUV: Creating refresh task")
        return self.post("refresh/tasks/", json={"access_token": access_token})

    def get_refresh_task(self, task_id: str) -> dict:
        logging.info("TRUV: Getting refresh task %s", task_id)
        return self.get(f"refresh/tasks/{task_id}/")

    # --- Admin ---

    def get_employee_directory_by_token(self, access_token: str) -> dict:
        logging.info("TRUV: Getting employee directory")
        return self.post("link/reports/admin/", json={"access_token": access_token})

    def request_payroll_report(self, access_token: str, start_date: str, end_date: str) -> dict:
        logging.info("TRUV: Requesting payroll report")
        return self.post(
            "administrators/payrolls/",
            json={"access_token": access_token, "start_date": start_date, "end_date": end_date},
        )

    def get_payroll_report_by_id(self, report_id: str) -> dict:
        logging.info("TRUV: Getting payroll report %s", report_id)
        return self.get(f"administrators/payrolls/{report_id}")
